src/features/payment/components/PaymentChoice/PaymentChoice.tsx

```diff
@@ -1,92 +1,147 @@
 import { DisabledPaymentMethods, MethodComponentList } from "./PaymentMethod";
+import { getNormalizedMethods } from "./utils";
 
-const shouldBeFirst = (method: PaymentInstrumentsType) =>
-  method.paymentTypeCode === PaymentCodeTypeEnum.CP;
-
-const sortMethods = (a: PaymentInstrumentsType, b: PaymentInstrumentsType) => {
-  if (shouldBeFirst(a)) {
-    return -1;
-  } else if (shouldBeFirst(b)) {
-    return 1;
+const callRecaptcha = async (recaptchaInstance: ReCAPTCHA, reset = false) => {
+  if (reset) {
+    void recaptchaInstance.reset();
   }
-  return a.description.localeCompare(b.description);
-};
-
-const getNormalizedMethods = (
-  paymentInstruments: Array<PaymentInstrumentsType>
-) => {
-  const { methods, duplicatedMethods } = paymentInstruments.reduce<{
-    foundTypes: Array<PaymentCodeType>;
-    methods: Array<PaymentInstrumentsType>;
-    duplicatedMethods: Array<PaymentInstrumentsType>;
-  }>(
-    (acc, method) =>
-      acc.foundTypes.includes(method.paymentTypeCode)
-        ? { ...acc, duplicatedMethods: acc.duplicatedMethods.concat(method) }
-        : {
-            ...acc,
-            methods: acc.methods.concat(method),
-            foundTypes: acc.foundTypes.concat(method.paymentTypeCode),
-          },
-    {
-      foundTypes: [],
-      methods: [],
-      duplicatedMethods: [],
-    }
-  );
-
-  const { enabled, disabled } = methods.reduce<{
-    enabled: Array<PaymentInstrumentsType>;
-    disabled: Array<PaymentInstrumentsType>;
-  }>(
-    (acc, method) =>
-      method.status === PaymentMethodStatusEnum.ENABLED
-        ? { ...acc, enabled: acc.enabled.concat(method) }
-        : { ...acc, disabled: acc.disabled.concat(method) },
-    { disabled: [], enabled: [] }
+  const recaptchaResponse = await recaptchaInstance.executeAsync();
+  return pipe(
+    recaptchaResponse,
+    O.fromNullable,
+    O.getOrElse(() => "")
   );
-  return {
-    enabled: enabled.slice().sort(sortMethods),
-    disabled: disabled.slice().sort(sortMethods),
-    duplicatedMethods,
-  };
 };
 
 export function PaymentChoice(props: {
   amount: number;
   paymentInstruments: Array<PaymentInstrumentsType>;
   loading?: boolean;
 }) {
-  const handleClickOnMethod = React.useCallback(
-    (paymentTypeCode: PaymentCodeType, paymentMethodId: string) => {
-      const route: string = PaymentMethodRoutes[paymentTypeCode]?.route;
-      setSessionItem(SessionItems.paymentMethod, {
-        paymentMethodId,
-        paymentTypeCode,
-      });
+  const ref = React.useRef<ReCAPTCHA>(null);
+  const [loading, setLoading] = React.useState(true);
+
+  const navigate = useNavigate();
+  const dispatch = useAppDispatch();
+
+  React.useEffect(() => {
+    if (ref.current) {
+      setLoading(false);
+    }
+  }, [ref.current]);
+
+  const onError = () => {
+    setLoading(false);
+    ref.current?.reset();
+  };
+
+  const getFees = (onSuccess: () => void) =>
```

Comment:
This is a "copy" of a similar function in the creditCard part. Please try to optimize with a single utility function avoiding duplications

Reply:
I'have addressed this issue on this [commit](https://github.com/pagopa/pagopa-checkout-fe/pull/318/commits/0d9b6eeb473099c2469a066a02f6f18cbe71fb96)